packages/block-editor/src/store/reducer.js

```diff
@@ -1509,11 +1509,20 @@ export function preferences( state = PREFERENCES_DEFAULTS, action ) {
 		case 'INSERT_BLOCKS':
 		case 'REPLACE_BLOCKS':
 			return action.blocks.reduce( ( prevState, block ) => {
-				let id = block.name;
-				const insert = { name: block.name };
+				const { attributes, name: blockName } = block;
+				const variations = getBlockVariations( blockName );
```

Comment:
The truth is I have mixed feelings about this as well. In general I had mixed feelings as well when I saw the similar function about Template Part, but it makes sense to me to hide the implementation detail of what is a reusable block etc..

In this case though in this reducer I think it's better to remove `isReusableBlock` for start.

>On principle, reducers should only contain logic that uses the previous state and action data to compute a new state

I mostly agree on this approach but I don't think is hard rule. A balance could be found based on the use case. They have a whole section about this in `Redux` [here](https://redux.js.org/faq/code-structure/#how-should-i-split-my-logic-between-reducers-and-action-creators-where-should-my-business-logic-go).

Regarding `getBlockVariations` we could move them to the actions `INSERT_BLOCKS` and `REPLACE_BLOCKS` but this would mean more code changes (here is at the sole reducer that needs this info) without an obvious benefit to me. Of course I'm still building my experience with Redux so I'm really open to feedback and thoughts.

Reply:
Speaking of docs, I'm referring to these guidelines:
https://redux.js.org/faq/reducers#reducers

> [Middleware with async logic](https://redux.js.org/tutorials/fundamentals/part-4-store#middleware) such as redux-thunk have access to the entire state through getState(). An action creator can retrieve additional data from the state and put it in an action so that each reducer has enough information to update its own state slice.

So to be clear what I meant here. In my opinion, it shouldn't be part of the regular action, but rather a side effect on in insert/replace block action that triggers another action like `INSERT_USAGE_INCREMENT` that passes the dynamically computed identifier that is used to update settings.

Reply:
Oh.. I understand you point now and it seems clean and separates concerns better. I will merge this though and look at this in a follow up.